spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizer.java

```diff
@@ -236,6 +242,14 @@ public Integer doInSession(Session<F> session) throws IOException {
 					F[] files = session.list(remoteDirectory);
 					if (!ObjectUtils.isEmpty(files)) {
 						List<F> filteredFiles = filterFiles(files);
+						if (maxFetchSize >= 0 && filteredFiles.size() > maxFetchSize) {
+							rollbackFromFileToListEnd(filteredFiles, filteredFiles.get(maxFetchSize));
+							List<F> newList = new ArrayList<>(maxFetchSize);
+							for (int i = 0; i < maxFetchSize; i++) {
+								newList.add(filteredFiles.get(i));
+							}
+							filteredFiles = newList;
```

Comment:
Hm... Generally I agree, but that would be possible if `maxFetchSize` can be changed at runtime from poll to poll. But that is just a setter option, so there might be potential race condition if we don't stop poller before changing that `maxFetch` option.